lib/i18n/backend/lazy_load.rb

```diff
@@ -0,0 +1,92 @@
+    # performance incentive for applications which have many translations for
+    # many different locales, but only refer to a single locales' translations
+    # at a time (ex. A Rails workload).  The implementation identifies which
+    # translation files from the load path belong to the current locale by
+    # pattern matching against their path name.
+    #
+    # Specifically, a translation file is considered to belong to a locale if:
+    # a) the filename is in the I18n load path
+    # b) the filename ends in a supported extension (ie. .yml, .json, .po, .rb)
+    # c1) the filename starts with the locale identifier
+    #    OR
+    # c2) the path contains a component equal to the locale identifier
+    #
+    # Examples:
+    # Valid files that will be selected by this backend:
+    #
+    # "files/locales/en-translation.yml" (Selected for locale "en")
+    # "files/locales/fr/translation.po"  (Selected for locale "fr")
+    #
+    # Invalid files that won't be selected by this backend:
+    #
+    # "files/locales/translation-file"
+    # "files/locales/en-translation.unsupported"
+    # "files/locales/french/translation.yml"
+    #
+    # The implementation uses this assumption to defer the loading of
+    # translation files until the current locale actually requires them.
+    #
+    # To use the LazyLoad backend instantiate it and set it to the I18n module.
+    # You can configure lazy loaded backends through the initializer or backends
+    # accessor:
+    #
+    #   # configure I18n backend to use LazyLoad Backend
+    #   I18n.backend = I18n::Backend::LazyLoad.new
+    #
+    class LazyLoad < Simple
+      # Returns whether the current locale is initialized.
+      def initialized?
+        initialized_locales[I18n.locale]
+      end
+
+      # Clean up translations and uninitialize all locales.
+      def reload!
+        @initialized_locales = nil
+        @translations = nil
+      end
+
+      protected
+
+      # Load translations from files that belong to the current locale.
+      def init_translations
+        load_translations(filenames_for_current_locale)
+        initialized_locales[I18n.locale] = true
+      end
+
+      def initialized_locales
+        @initialized_locales ||= Hash.new(false)
+      end
+
+      # Select all files from I18n load path that belong to current locale.
+      # These files must start with the locale identifier (ie. "en", "fr").
+      # or contain a path component equal to the locale identifier (ie. /locales/en/translation.yml)
+      def filenames_for_current_locale
+        I18n.load_path.flatten.select do |path|
+          (basename_starts_with_locale?(path) || path_contains_locale_identifier?(path)) &&
+          supported_extension?(path)
+        end
+      end
+
+      SUPPORTED_EXTENSIONS = [".yml", ".yaml", ".po", ".json", ".rb"].freeze
+
+      def supported_extension?(path)
+        SUPPORTED_EXTENSIONS.include?(File.extname(path))
```

Comment:
Not sure how much of a hotpath this is, but my hunch says this could be faster:

```suggestion
        path.end_with?(*SUPPORTED_EXTENSIONS)
```

Did a quick bench:

```ruby
# frozen_string_literal: true

require 'benchmark/ips'

SUPPORTED_EXTENSIONS = [".yml", ".yaml", ".po", ".json", ".rb"].freeze

Benchmark.ips do |x|
  x.report("end_with?") { "/Users/byroot/src/github.com/Shopify/shopify/config/database.yml".end_with?(*SUPPORTED_EXTENSIONS) }
  x.report("include? ext") { SUPPORTED_EXTENSIONS.include?(File.extname("/Users/byroot/src/github.com/Shopify/shopify/config/database.yml")) }
  x.compare!
end
```

```
Warming up --------------------------------------
           end_with?     1.128M i/100ms
        include? ext   181.313k i/100ms
Calculating -------------------------------------
           end_with?     11.262M (± 0.2%) i/s -     56.396M in   5.007837s
        include? ext      1.811M (± 0.3%) i/s -      9.066M in   5.006622s

Comparison:
           end_with?: 11261609.5 i/s
        include? ext:  1810746.8 i/s - 6.22x  (± 0.00) slower
```

Reply:
Thanks for the benchmark, changed 👍